src/Thruway/Procedure.php

```diff
@@ -249,44 +256,93 @@ public function processCall(Session $session, Call $call)
     public function processQueue()
     {
         if (!$this->getAllowMultipleRegistrations()) {
-            throw new \Exception("queuing only allowed when there are multiple registrations");
+            throw new \Exception("Queuing only allowed when there are multiple registrations");
         }
 
         // find the best candidate
         while ($this->callQueue->count() > 0) {
-            $congestion = true;
-
-            /* @var $bestRegistration \Thruway\Registration */
-            $bestRegistration = $this->registrations[0];
-            /* @var $registration \Thruway\Registration */
-            foreach ($this->registrations as $registration) {
-                if ($registration->getSession()->getPendingCallCount() == 0) {
-                    $bestRegistration = $registration;
-                    $congestion       = false;
-                    break;
-                }
-                if ($registration->getSession()->getPendingCallCount() <
-                    $bestRegistration->getSession()->getPendingCallCount()
-                ) {
-                    $bestRegistration = $registration;
-                }
+            $registration = NULL;
+
+            if (strcasecmp($this->getInvokeType(), Registration::FIRST_REGISTRATION) === 0) {
+                $registration = $this->getNextFirstRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::LAST_REGISTRATION) === 0) {
+                $registration = $this->getNextLastRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::RANDOM_REGISTRATION) === 0) {
+                $registration = $this->getNextRandomRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::ROUNDROBIN_REGISTRATION) === 0) {
+                $registration = $this->getNextRoundRobinRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::THRUWAY_REGISTRATION) === 0) {
+                $registration = $this->getNextThruwayRegistration();
             }
 
-            if ($congestion) {
-                // there is congestion
-                $bestRegistration->getSession()->getRealm()->publishMeta('thruway.metaevent.procedure.congestion',
-                    [
-                        ["name" => $this->getProcedureName()]
-                    ]
-                );
+            if ($registration !== NULL) {
+                $call = $this->callQueue->dequeue();
+                $registration->processCall($call);
+            }
```

Comment:
Yes, that looks to be the case, but even so, we can't have the router block. Just breaking from the while loop if it is null would be sufficient to keep the thruway type working.

The entire queuing idea only needs to be done for the thruway type as the other invocation types do not queue, as they always send the call to one of the registrations.